Model/ModelManager.php

```diff
@@ -376,7 +386,7 @@ public function getNormalizedIdentifier($entity)
         }
 
         // the entities is not managed
-        if (!$entity /*|| !$this->getEntityManager($entity)->getUnitOfWork()->isInIdentityMap($entity) // commented for perfomance concern */) {
+        if (!$entity || !$this->getEntityManager($entity)->getUnitOfWork()->isInIdentityMap($entity)) {
```

Comment:
My point here is that I think that `UnitOfWork::isInIdentityMap()` is too unreliable. `UnitOfWork::getEntityState()` will give you one of four values, and from this you can tell if the object is persisted or not. In my case it returns `STATE_MANAGED` before the entity leaves the UOW and `STATE_DETACHED` afterwards. And both states identify that the entity is persisted in the database. Any new (unpersisted and unmanaged) object should return `STATE_NEW` and anything pending delete should return `STATE_REMOVED`.

If you're trying to test if an object has been persisted already then I think that checking for `STATE_MANAGED || STATE_DETACHED` would be more robust than simply using `isInIdentityMap()`.

However, as I said before, I don't really understand what you're trying to do in the case where identifiers are set manually, so perhaps I've missed something? However, if you just want this method to return null in cases where the entity in hand is unmanaged/unpersisted then you can do something like:

    if (!$entity) {
        return null;
    }

    $entityState = $this->getEntityManager($entity)->getUnitOfWork()->getEntityState($entity);
    if (UnitOfWork::STATE_NEW == $entityState || UnitOfWork::STATE_REMOVED == $entityState) {
        return null;
    }

Would that provide the required functionality in cases where IDs are set manually?